### ISSUE TITLE
Make thread safety analysis warnings be able to look through moves

### ISSUE THREAD

**Issue by Weverything** [llvm/llvm-project#59662]
Using thread safety annotations with smart pointers:
```
#include <absl/synchronization/mutex.h>
#include <memory>

struct A {
    mutable
    absl::Mutex m;
};

void bar(std::unique_ptr<A> a)
ABSL_UNLOCK_FUNCTION(a->m) {
  a->m.Unlock();
}

void foo(std::unique_ptr<A> a)
ABSL_LOCKS_EXCLUDED(a->m) {
  a->m.Lock();
  bar(std::move(a));
}
```

Gives the following false positive:
```
<source>:18:3: warning: releasing mutex '#undefined->m' that was not held [-Wthread-safety-analysis]
  bar(std::move(a));
  ^
<source>:19:1: warning: mutex 'a->m' is still held at the end of function [-Wthread-safety-analysis]
}
^
<source>:17:8: note: mutex acquired here
  a->m.Lock();
       ^
2 warnings generated.
```

The analysis sees in function `foo` that `a->m` is locked, but since `bar` is called with moved value, it thinks that `std::move(a)->m` is unlocked.  Looking through moves to equate `a` and `std::move(a)` would fix this issue.

Example in godbolt: https://godbolt.org/z/baGWGvGEz

**Comment by aaronpuchert** [comment]
There are two issues here:
* The analysis doesn't look into function calls. This limitation is [intentional](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#no-inlining). However, inlining `std::move(a)` as `static_cast<std::unique_ptr<A>&&>(a)` should work. When it comes to understanding standard library functions, I'm not sure where to draw the line. Should we only accept `std::move`, or any `move` with one argument?
* Since `bar` takes the `unique_ptr` by value, we have an additional `CXXConstructExpr` around the argument. So technically we have a new value here, and we [don't really track aliases](https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#no-alias-analysis). We would need to understand special member functions and assume that they don't do anything strange. It could be tricky to collect knowledge such as “this `CXXConstructExpr` has the same value as `a`”. Or should we simply look past copy/move constructor calls in a `CXXBindTemporaryExpr`?

If you need a workaround, this is accepted:
```c++
void bar(std::unique_ptr<A>&& a) // Note the added &&.
ABSL_UNLOCK_FUNCTION(a->m) {
  a->m.Unlock();
}

void foo(std::unique_ptr<A> a)
ABSL_LOCKS_EXCLUDED(a->m) {
  a->m.Lock();
  bar(static_cast<std::unique_ptr<A>&&>(a)); // Inlined std::move.
}
```
